src/app/EventManagement.cpp

```diff
@@ -0,0 +1,903 @@
+    mpEventBuffer->mAppData               = nullptr;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR EventManagement::BlitEvent(EventLoadOutContext * apContext, EventLoggingDelegate * apDelegate,
+                                      const EventOptions * apOptions)
+{
+
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    TLVWriter checkpoint = apContext->mWriter;
+    chip::TLV::TLVType dataContainerType;
+    chip::app::EventDataElement::Builder eventDataElementBuilder;
+    chip::app::EventPath::Builder eventPathBuilder;
+
+    VerifyOrExit(apContext->mCurrentEventNumber >= apContext->mStartingEventNumber,
+                 /* no-op: don't write event, but advance current event Number */);
+
+    VerifyOrExit(apOptions != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(apOptions->mTimestamp.mType != Timestamp::Type::kInvalid, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    eventDataElementBuilder.Init(&(apContext->mWriter));
+    eventPathBuilder = eventDataElementBuilder.CreateEventPathBuilder();
+    SuccessOrExit(eventPathBuilder.GetError());
+
+    // TODO: Revisit NodeId since the the encoding spec and the IM seem to disagree on how this stuff works
+    eventPathBuilder.NodeId(apOptions->mpEventSchema->mNodeId)
+        .EndpointId(apOptions->mpEventSchema->mEndpointId)
+        .ClusterId(apOptions->mpEventSchema->mClusterId)
+        .EventId(apOptions->mpEventSchema->mEventId)
+        .EndOfEventPath();
+    SuccessOrExit(eventPathBuilder.GetError());
+
+    eventDataElementBuilder.PriorityLevel(static_cast<uint8_t>(apContext->mPriority));
+
+    // if mFirst, record absolute time
+    if (apContext->mFirst)
+    {
+        if (apOptions->mTimestamp.mType == Timestamp::Type::kSystem)
+        {
+            eventDataElementBuilder.Number(apContext->mCurrentEventNumber);
+            eventDataElementBuilder.SystemTimestamp(apOptions->mTimestamp.mValue);
+        }
+    }
+    // else record delta
+    else
+    {
+        if (apOptions->mTimestamp.mType == Timestamp::Type::kSystem)
+        {
+            uint64_t deltatime = apOptions->mTimestamp.mValue - apContext->mCurrentSystemTime.mValue;
+            eventDataElementBuilder.DeltaSystemTimestamp(deltatime);
+        }
+    }
+
+    SuccessOrExit(eventDataElementBuilder.GetError());
+
+    err = apContext->mWriter.StartContainer(ContextTag(chip::app::EventDataElement::kCsTag_Data), chip::TLV::kTLVType_Structure,
+                                            dataContainerType);
+    SuccessOrExit(err);
+    // Callback to write the EventData
+    err = apDelegate->WriteEvent(apContext->mWriter);
+    SuccessOrExit(err);
+
+    err = apContext->mWriter.EndContainer(dataContainerType);
+    SuccessOrExit(err);
+
+    eventDataElementBuilder.EndOfEventDataElement();
+    SuccessOrExit(eventDataElementBuilder.GetError());
+
+    err = apContext->mWriter.Finalize();
+    SuccessOrExit(err);
+
+    // only update mFirst if an event was successfully written.
+    if (apContext->mFirst)
```

Comment:
the if does not seem needed. can we just do `apContext->mFirst = false` ?

Reply:
yep, done, thanks